examples/cosmos/ics20/bank.qnt

```diff
@@ -0,0 +1,240 @@
+// -*- mode: Bluespec; -*-
+
+/**
+ * A bank module for ICS-020 (Fungible Token Transfer).
+ *
+ * ICS-020 assumes a high-level "bank module" capable of minting, burning, and
+ * transfering tokens. This specification is held as general as possible,
+ * providing only the assumptions required by ICS-020.
+ *
+ * For a possible implementation, consider the Cosmos SDK bank module:
+ * https://docs.cosmos.network/v0.47/modules/bank/
+ *
+ * Thomas Pani, Informal Systems, 2023
+ */
+
+module bank {
+
+  /* ***************************************************************************
+   * TYPES
+   * **************************************************************************/
+
+  type Address = str
+  type Denomination = str
+  type Amount = int
+
+  type TokenBalances = Denomination -> Amount
+  type Accounts = Address -> TokenBalances
+
+  /* ***************************************************************************
+   * FUNCTIONAL LAYER, manipulating balances
+   * **************************************************************************/
+
+  // Get balances of all denominations in the given account `addr`.
+  pure def getBalances(accounts: Accounts, addr: Address) : TokenBalances = {
+    if (accounts.keys().contains(addr)) {
+      accounts.get(addr)
+    } else {
+      Map()
+    }
+  }
+
+  // Get the amount of tokens of denomination `denom` in `balances`.
+  pure def getBalance(balances: TokenBalances, denom: Denomination): Amount = {
+    if (balances.keys().contains(denom)) balances.get(denom) else 0
+  }
+
+  /* Return `accounts`, where `address` is updated to hold
+   * `currentAmount + amount` tokens of denomination `denom`.
+   * 
+   * `currentAmount` refers to the number of tokens of denomination `denom` held
+   * by `address` in `accounts`.
+   *
+   * If there currently is no account under `address`, return accounts including
+   * `address` with exactly `amount`-many `denom` tokens.
+   */
+  pure def manipulateAccount(accounts: Accounts, address: Address,
+                             denom: Denomination, amount: Amount) : Accounts = {
+    val accountCoins = accounts.getBalances(address)
+    val newCoins = accountCoins.put(denom, accountCoins.getBalance(denom) + amount)
+    accounts.put(address, newCoins)
+  }
+
+  /* Return `accounts`, where `amount`-many tokens of denomination `denom`
+   * vanished out of `sender`'s account into thin air.
+   *
+   * This intentionally does NOT check whether `amount > 0`.
+   */
+  pure def BurnCoins(accounts: Accounts, sender: Address,
```

Comment:
```suggestion
  pure def burnCoins(accounts: Accounts, sender: Address,
```

Let's just agree on whether we make the first letter capital or not. I would prefer to keep it small (non-capital).

Reply:
Since we have no access modifiers, I made the "public" ones uppercase and the "private" ones lowercase. I think it makes it easier to see what the actual interface is. Do you have other suggestions?

Reply:
The pseudo-go in the English spec uses this casing, so I'd rather keep it in sync.